jl4-lsp/app/LSP/L4/Handlers.hs

```diff
@@ -333,19 +333,58 @@ handlers recorder =
         pure (Right (InL locs))
 
     -- custom requests
-    , requestHandler (SMethod_CustomMethod (Proxy @Ladder.EvalAppMethodName)) $ \_ide params -> do
+    , requestHandler (SMethod_CustomMethod (Proxy @Ladder.EvalAppMethodName)) $ \ide params -> do
         let methodName = getMethodName (Proxy @Ladder.EvalAppMethodName)
+        
         case Aeson.fromJSON params :: Aeson.Result EvalAppRequestParams of
-          Aeson.Success evalParams -> do
-            logWith recorder Debug $ LogReceivedCustomRequest evalParams.verDocId._uri methodName
-            -- Haven't implemented the actual call to eval yet; just returning null for now
-            pure $ Right Aeson.Null
           Aeson.Error err -> do
             pure $ Left $ TResponseError
               { _code = InR ErrorCodes_InvalidRequest
               , _message = "Invalid params for " <> methodName <> ": " <> Text.pack err
               , _xdata = Nothing
               }
+          
+          -- eval params valid
+          Aeson.Success evalParams -> do
+            logWith recorder Debug $ LogReceivedCustomRequest evalParams.verDocId._uri methodName
+            
+            mtcRes <- liftIO $ runAction "l4/evalApp" ide $ use TypeCheck $ toNormalizedUri evalParams.verDocId._uri
+            mRecentViz <- liftIO $ atomically $ getMostRecentVisualisation ide
+            
+            runExceptT $ case (mtcRes, mRecentViz) of
+              -- impossible
+              (Nothing, _) -> defaultResponseError $ "Failed to get type check for " <> Text.show evalParams.verDocId._uri
+              (_, Nothing) -> defaultResponseError $ "No recent visualisation found, when trying to handle " <> methodName <> ". This case should be impossible."
+              
+              -- Another approach would be to make a new virtual file with the existing module + our directive,
+              -- and use oneshotL4ActionAndErrors on that new file
+              
+              {- We require that the client's verTxtDocId matches the server's.
+                 Note that the client will have already received 
+                 the verTxtDocId in the original payload for the 'please render this VizExpr' request -}
+              (Just tcRes, Just recentViz@(Ladder.getVizConfig . (.vizState) -> vizConfig))
```

Comment:
please don't use viewpatterns if you don't actually match any patterns

Reply:
Good point thank you. You've pointed this out before, sorry for not having internalized this